stan/math/prim/mat/fun/cholesky_decompose.hpp

```diff
@@ -24,11 +24,15 @@ namespace math {
  * <p>\f$A = L \times L^T\f$.
  * @param m Symmetrix matrix.
  * @return Square root of matrix.
+ * @note Because OpenCL only works on doubles there are two
+ * <code>cholesky_decompose</code> functions. One that works on doubles
+ * (this one) and another that works on all other types.
  * @throw std::domain_error if m is not a symmetric matrix or
  *   if m is not positive definite (if m has more than 0 elements)
  */
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_decompose(
+template <typename T, typename std::enable_if_t<std::is_same<T, double>::value,
+                                                T>* = nullptr>
+inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_decompose(
```

Comment:
So... this is saying we want to activate this specialization only for doubles, right?

Why don't we just specialize on double? e.g.
```
inline Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> cholesky_decompose(
        const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& m) {
```

Reply:
I think that makes a lot more sense.

Though the templating here is gross, I wanted them to have similar templates for consistency. But yeah the above is a lot nicer and does the job

Reply:
The other one also becomes simpler, assuming my mental model doesn't have any holes in it:

```
template <typename T>
inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_decompose(
    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
```

Reply:
I think the first one might need a `template <>` above it, oops.

This page has a lot more detail on how template specialization works: https://isocpp.org/wiki/faq/templates#template-specialization

Reply:
Oh interesting! My assumption was that the template

```cpp
template <typename T>
inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_decompose(
    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
```

Would overwrite the doubles specific template. But your above makes that seem false. Ty! I'll read that more tonight and try out your approach.

Reply:
If you dont mind, I can make the changes and commit rn.

Reply:
I messed up the order in the first try, the double specialization needs to be written after the general template function.